docs/spec/concepts.rst

```diff
@@ -3,97 +3,216 @@
-          return x * 2
-
-Since the subclasses of ``Enum`` cannot be further subclassed,
-the type of variable ``x`` can be statically inferred in all branches
-of the above example. The same approach is applicable if more than one
-singleton object is needed: one can use an enumeration that has more than
-one value::
-
-  class Reason(Enum):
-      timeout = 1
-      error = 2
-
-  def process(response: str | Reason = '') -> str:
-      if response is Reason.timeout:
-          return 'TIMEOUT'
-      elif response is Reason.error:
-          return 'ERROR'
-      else:
-          # response can be only str, all other possible values exhausted
-          return 'PROCESSED: ' + response
+Static, dynamic, and gradual typing
+-----------------------------------
+
+A **statically-typed** programming language runs a type-checker before running
+a program. The program is required to be well-typed according to the language's
+type system. The type system assigns a type to all expressions in the language
+and verifies that their uses obey the typing rules. Normally, a program that is
+not well-typed (i.e., one that contains a type error) will not run. Java and
+C++ are examples of statically-typed object-oriented languages.
+
+A **dynamically-typed** programming language does not run a type-checker before
+running a program. Instead, it will check the types of values before performing
+operations on them at runtime. This is not to say that the language is
+"untyped". Values at runtime have a type and their uses obey typing rules. Not
+every operation will be checked, but certain primitive operations in the
+language such as attribute access or arithmetic will be. Python was
+historically a dynamically-typed language.
+
+**Gradual typing** is the name for a specific way to combine static and dynamic
+typing. Type-annotated Python allows mixing static and dynamic checking at a
+fine level of granularity. Individual variables, parameters, and returns can
+optionally be given a static type. Data structures such as objects can have a
+mix of static and dynamic-only checking. For example, a dictionary could be
+annotated to have static checking of the key type but only dynamic checking of
+the value type.
+
+In gradual typing, a dynamically typed term is indicated by a special "unknown"
+or "dynamic" type.  In Python, the unknown type is spelled :ref:`Any`. It
+indicates to the static type checker that this term should not be subject to
+static checking.  The system should not signal a static type error for use of a
+term with type :ref:`Any`.  Instead, the term's type will be dynamically
+checked, according to the Python runtime's usual checks on the types of runtime
+values.
+
+This specification describes a gradual type system for Python.
+
+Static, dynamic, and gradual types
+----------------------------------
+
+We will refer to types that do not contain :ref:`Any` as a sub-part as **static
```

Comment:
According to the definitions I currently give in this PR, only `Any` is "the dynamic type." All types (including `Any`, `list[Any]`, and e.g. `int`) are "gradual types." Only types that don't involve `Any` are "static types." (The category "gradual type" encompasses the category "static type.")

It is important to make this distinction somehow because it is fundamental to the definitions of subtyping, materialization, and consistency. Only static types represent a specific set of Python objects and behave according to a set-theoretic view of subtyping. Assignability (or consistent subtyping) requires both materialization of gradual types to static types and the subtype relation on static types.

Using the term "dynamic type" for Any is very well established in the literature (in many treatments it is actually called `dynamic` or `Dyn`). I think it's also intuitive because Any does defer type checking to the runtime, which is precisely what "dynamic typing" or "dynamically typed" means. So I don't see a good case for avoiding the use of "dynamic type" to refer to Any.

If Any is "the dynamic type" it seems logical to name types not involving Any as "static types." This is also reasonable because these are the only types that appear in fully static (i.e. non-gradual) type systems. Kevin's doc used the term "fully static types" for these; I'd be fine with that term if it seems clearer. Also open to considering other options if there are suggestions.

Reply:
If `Any` is the only example of "dynamic type" then why does it need a separate term? Let's just call it `Any` then. 

TypeScript has an `any` type that is analogous to the one in Python's type system, but I don't think you'll find any reference to "dynamic type" in TypeScript's specification.

I'm good with the term "gradual type" to mean "any type that incorporates `Any`, `...` or similar gradual typing affordances. 

We could then eliminate the term "static type", which I think is really confusing given that these are all "static type annotations". The term "non-gradual type" could be used in the few places where the spec needs to make a clear distinction between gradual and non-gradual types.

Reply:
I find it hard to write this spec without a term for "types that don't contain Any", because that's a really important concept -- only these types represent a single set of objects, and participate directly in subtyping. I really don't like the way it reads to only use a negative description (e.g. "non-gradual") of such types. Also, it reads much better to define "gradual type" as including all types, otherwise we end up with a lot of "gradual and non-gradual types" all over the place. (Could maybe just be "types" but I think this is not clear enough.)

I switched from "static type" to "fully static type" in the hopes that this might help address your concern that "static type" is confusing.

I still think that it is highly valuable to describe the materialization preorder as a continuum from "more dynamic" types (which contain more non-statically-known components) to "more static" types (which contain fewer), all the way to "fully static" types (which contain no non-statically-known components.) And I still feel that the terms "dynamic" and "static" are quite sensible in this context, as the latter are more "statically known" and "statically checked", while the former are less statically known, and thus (to some degree) more reliant on the dynamic checking which applies to all values at runtime.

Will be grateful for your review of the latest version.